pkg/operator/operator.go

```diff
@@ -743,3 +786,337 @@ func (o *Operator) Config(ctx context.Context, key string) (*manifests.Config, e
+		err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			node, err := o.client.GetNode(ctx, node.Name)
+			if err != nil {
+				return err
+			}
+			delete(node.Annotations, cordonAnnotation)
+			_, err = o.client.UpdateNode(ctx, node)
+			return err
+		})
+		if err != nil {
+			return false, err
+		}
+	}
+
+	return true, nil
+}
+
+func (o *Operator) rebalanceWorkloads(ctx context.Context, namespace string, sel map[string]string) (bool, error) {
+	podList, err := o.client.ListPods(ctx, namespace, metav1.ListOptions{LabelSelector: labels.FormatLabels(sel)})
+	if err != nil {
+		return false, err
+	}
+
+	if len(podList.Items) <= 1 {
+		return false, nil
+	}
+
+	var (
+		pvcsToDelete []v1.PersistentVolumeClaim
+		podsToDelete []v1.Pod
+	)
+	for _, pod := range podList.Items {
+		for _, vol := range pod.Spec.Volumes {
+			if vol.PersistentVolumeClaim == nil {
+				continue
+			}
+			pvc, err := o.client.GetPersistentVolumeClaim(ctx, pod.Namespace, vol.PersistentVolumeClaim.ClaimName)
+			if apierrors.IsNotFound(err) {
+				break
+			}
+			if err != nil {
+				return false, err
+			}
+
+			dropPVC, ok := pvc.Annotations[dropPVCAnnotation]
+			if !ok {
+				break
+			}
+			if dropPVC == "yes" {
+				pvcsToDelete = append(pvcsToDelete, *pvc)
+				podsToDelete = append(podsToDelete, pod)
+			}
+		}
+	}
+
+	var workloadRebalanced int
+	for i, pod := range podsToDelete {
+		pvc := pvcsToDelete[i]
+		// Guard from deleting all PVCs to prevent complete data loss.
+		if workloadRebalanced == len(podList.Items)-1 {
+			// Remove the annotation so that the PVC doesn't get deleted in future cycles.
+			klog.V(4).Infof("Removing annotation %s from PersistentVolumeClaim %s/%s to avoid needlessly deleting all PVCs to rebalance a workload.", dropPVCAnnotation, pvc.Namespace, pvc.Name)
+			err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				pvc, err := o.client.GetPersistentVolumeClaim(ctx, pvc.Namespace, pvc.Name)
+				if err != nil {
+					return err
+				}
+				delete(pvc.Annotations, dropPVCAnnotation)
+				_, err = o.client.UpdatePersistentVolumeClaim(ctx, pvc)
+				return err
+			})
+			if err != nil {
+				return false, err
+			}
+			break
+		}
+
+		klog.V(4).Infof("Rebalancing pod %s/%s.", pod.Namespace, pod.Name)
+		rebalanced, err := o.rebalanceWorkload(ctx, &pod, &pvc)
+		if err != nil {
+			return false, err
+		}
+
+		if rebalanced {
+			workloadRebalanced++
+		}
+	}
+
+	return workloadRebalanced > 0, nil
+}
+
+// workloadCorrectlyBalanced returns whether the selected pods are balanced
+// across different nodes ensuring proper high-availability.
+// If the pods don't use persistent storage, it will always return true.
+func (o *Operator) workloadCorrectlyBalanced(ctx context.Context, namespace string, sel map[string]string) (bool, error) {
+	podList, err := o.client.ListPods(ctx, namespace, metav1.ListOptions{LabelSelector: labels.FormatLabels(sel)})
```

Comment:
same question here about keeping only `Running` pods

Reply:
I think that would make sense in this case since we want to check that the running pods are located on different nodes.